File: client/lib/src/envelope/envelope_verifier.dart
import 'dart:typed_data';

import 'package:sodium/sodium.dart';

import '../crypto/ed25519.dart';
import '../crypto/sha256.dart';
import 'envelope_codec.dart';
import 'envelope_rejection.dart';
import 'issuer_key_resolver.dart';
import 'shared_envelope.dart';

/// Verifies a [SharedEnvelope] against the issuer directory.
///
/// Rejection order (mirrors Java):
/// 1. Parse / canonical check → [MalformedCbor]
/// 2. Version check           → [UnsupportedVersion]
/// 3. Key lookup + validity   → [IssuerKeyNotActiveAt]
/// 4. Signature check         → [SignatureInvalid]
/// 5. Ciphertext hash check   → [CiphertextHashMismatch]
class EnvelopeVerifier {
  final EnvelopeCodec _codec;
  final IssuerKeyResolver _keyResolver;
  final Sodium _sodium;

  EnvelopeVerifier(this._sodium, this._keyResolver) : _codec = EnvelopeCodec();

  /// Verifies [envelopeBytes] and returns the parsed [SharedEnvelope].
  ///
  /// Throws a subtype of [EnvelopeRejection] on any verification failure.
  SharedEnvelope verify(Uint8List envelopeBytes) {
    final envelope = _codec.decode(envelopeBytes); // may throw MalformedCbor

    if (envelope.version != 1) {
      throw UnsupportedVersion(envelope.version);
    }

    final keyView = _keyResolver(
        envelope.issuerId, envelope.issuerSigningKeyId, envelope.createdAt);
    if (keyView == null) {
      throw IssuerKeyNotActiveAt(
          'No directory record for issuer key at createdAt=${envelope.createdAt}');
    }
    if (envelope.createdAt < keyView.validFrom ||
        envelope.createdAt >= keyView.validUntil) {
      throw IssuerKeyNotActiveAt(
          'createdAt=${envelope.createdAt} outside valid range '
          '[${keyView.validFrom}, ${keyView.validUntil})');
    }

    final signedBytes =
        _codec.signedBytesOf(envelope.withSignature(null));
    if (!Ed25519.verifyDetached(
        _sodium, keyView.publicKey, signedBytes, envelope.signature!)) {
      throw const SignatureInvalid();
    }

    final computedHash = Sha256.hash(envelope.ciphertext);
    if (!_bytesEqual(computedHash, envelope.ciphertextHash)) {
      throw const CiphertextHashMismatch();
    }

    return envelope;
  }

  static bool _bytesEqual(Uint8List a, Uint8List b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }
}

File: src/main/java/com/erikromson/datawallet/envelope/EnvelopeVerifier.java
package com.erikromson.datawallet.envelope;

import com.erikromson.datawallet.crypto.Ed25519;
import com.erikromson.datawallet.crypto.Sha256;

import java.util.Arrays;
import java.util.Optional;

public final class EnvelopeVerifier {

    private final EnvelopeCodec codec;
    private final IssuerKeyResolver keyResolver;

    public EnvelopeVerifier(EnvelopeCodec codec, IssuerKeyResolver keyResolver) {
        this.codec = codec;
        this.keyResolver = keyResolver;
    }

    public SharedEnvelope verify(byte[] envelopeBytes) {
        SharedEnvelope envelope = codec.decode(envelopeBytes);

        if (envelope.version() != 1) {
            throw new EnvelopeRejection.UnsupportedVersion(envelope.version());
        }

        Optional<DirectoryKeyView> keyView = keyResolver.resolve(
                envelope.issuerId(), envelope.issuerSigningKeyId(), envelope.createdAt()
        );
        if (keyView.isEmpty()) {
            throw new EnvelopeRejection.IssuerKeyNotActiveAt(
                    "No directory record for issuer key at createdAt=" + envelope.createdAt()
            );
        }
        DirectoryKeyView dk = keyView.get();
        if (envelope.createdAt() < dk.validFrom() || envelope.createdAt() >= dk.validUntil()) {
            throw new EnvelopeRejection.IssuerKeyNotActiveAt(
                    "createdAt=" + envelope.createdAt()
                            + " outside valid range [" + dk.validFrom() + ", " + dk.validUntil() + ")"
            );
        }

        byte[] signedBytes = codec.signedBytesOf(
                envelope.withSignature(null)
        );
        if (!Ed25519.verifyDetached(dk.publicKey(), signedBytes, envelope.signature())) {
            throw new EnvelopeRejection.SignatureInvalid();
        }

        byte[] computedHash = Sha256.hash(envelope.ciphertext());
        if (!Arrays.equals(computedHash, envelope.ciphertextHash())) {
            throw new EnvelopeRejection.CiphertextHashMismatch();
        }

        return envelope;
    }
}
